Providers are found in local project dir after tf init

Useurmind opened the pull request:
Hi,

I created a fix for this Issue:
https://github.com/nhurel/terraspec/issues/7

With terraform 0.13 `terraform init` will pull all providers into subfolders of the `.terraform/plugins` folder.
I search the providers there, extract their hostname and namespace and create the correct metadata for them.

I added a few tests to validate the behaviour.

The only remaining issue is: When executing main_test.go the test cannot remove the `.terraform` folder. Probably because a provider is still accessed or open or something like that.

nhurel commented:
Thanks for your PR.

Regarding the fix itself, I think what you did is good but it still doesn't support loading plugins from user home directory, as you raised in the original issue. If you could add the user directory to the scanned folders, that would be really great.

Regarding the test parts, i'm not comfortable with your approach. The test looks more as an integrration test rather than a unit test and it's doing too many things on the running host : creating directories, downloading 3rd party plugins from the internet, running `terraform init`, ... To my mind, this is too heavy for a unit test and moreover it wouldn't pass when tests are executed as github actions.
Nevertheless, I had a look at how `discovery.FindPlugins` works and I think that you could write a unit test that only tests the `BuildProviderResolver` function if you prepare a dedicated testdata directory with right directory names and empty files to mock downloaded plugins.  Does it makes sense ?

(Last point : I haven't added any test framework to the project so I'd appreciate if you could remove testify from the dependencies)

Let me know if I'm unclear or if you need help to finish your work.

Useurmind commented:
Yes I know that this PR does not implement a fix to the original problem. But the original problem is one of Terraform 0.12. This fix is for Terraform 0.13 which behaves completely different regarding provider resolution.

As Terraspec currently requires a Terraform init before it can work I designed this PR around that fact. Terraform does a lot of things to get the necessary plugins into place. I would not try to replicate that logic. But since version 0.13 it also pulls all the required plugins into the local folder I am scanning here during tf init.

The problem with just adding the home folder is that the home folder is not even the only location Terraform is looking into. Therefore it's better to let tf init gather the plugins into the local  `.terraform/plugins` folder and then continue from there.

Regarding the tests. If you prefer another assertion/test framework let me know. I don't know any other currently. But implementing the asserts by hand would be rather pointless.

I also know that the test are integration tests. But I also think that we need them for this functionality as we are integrating with Terraform provider resolution which is not part of the codebase of terraspec. I could try to make them work with GitHub actions if you like. This tools probably needs a lot of integration tests to stay stable in regards to Terraform compatibility.

Useurmind commented:
So I pretty much finished with the tasks on my list now.

- Removed testify asserts
- Moved integration tests out into separate package so they can be executed optionally
- Added simple unit tests
- Improved the detection logic
- Made sure tests run in windows and linux (wsl on my machine)

nhurel commented:
Many thanks for the hard work.

I've checked out your PR and ran the tests. The unit test work fine, including TestBuildProviderResolver but the 2 integration tests are failing (which demonstrates the need for integration tests, you're right :wink:   ). It looks like the aws provider is found but it fails to load the "nocorp" one.
I've also tried to : 
- go in test_project directory
- run `terraform init`
- run `go run ../../main.go` 
and it also complains about the nocorp plugin.

I'm testing from a linux laptop.
I can investigate on this and push on your branch if you want.

Apart from that, I have 2 comments on the code itself :
- The `parseProviderValues` will only work if terraform init was run with terraform >= 0.13. I'd like to keep the backward compatibility with terraform 0.12 as much as possible (I think the method could handle the 2 cases based on the length of the path from the .terraform directory)
- The integration tests would be even more "integration test" if they effectively run the terraspec binary rather than calling
 terraspec.ExecTerraspec : Calling the binary will avoid the need of moving everything from main.go in lib/terraspec.go + it could help to eg. add testing on the output of the program in the future.

Again, let me know if you want me to help on this.

Useurmind commented:
Both comments sound reasonable.

I will look into them today and report back how far I got and if I need help with it.

Useurmind commented:
I have finished a new version for this PR.

- Integration tests use terraspec executable
- Terraform 12 compatible
- Linux tests work

Just to give you a summary.
- Linux tests did not work because `terraform init` does not copy in linux but uses symlinks, `filepath.Walk` does not follow those so I replaced it with a function that does
- Another point worth mentioning is that for terraform 12 I present the underlying terraform 13 runtime integrated in terraspec with the provider namespace "hashicorp" and the default host "registry.terraform.io" because that is where terraform 13 tries to locate providers without explicit requirement in the `terraform` block. That seems a little bit hacky to me but currently it works.

nhurel commented on integration_tests/test_helper.go:
```diff
@@ -0,0 +1,348 @@
+		t.Fatalf("Could not unzip terraform download: %v", err)
+	}
+
+	return terraformPath, func() {}
```
the returned deleteTerraform function is alway empty ?

Useurmind replied:
I dont delete it anymore because I kind of "cache" them for all the tests. I will remove that return param.

nhurel commented on lib/providers.go:
```diff
@@ -62,16 +67,98 @@ func (m *MockDataSourceReader) UnmatchedCalls() []cty.Value {
+
+	if parts[partCount-3] == "plugins" {
+		// plugins folder was initialized by terraform <=0.12
+		return &addrs.Provider{
+			Hostname:  addrs.DefaultRegistryHost,
+			//Namespace: addrs.LegacyProviderNamespace,
+			// Fake this, because terraform 13 will look for this namespace by default
+			Namespace: "hashicorp",
+			Type:      provMeta.Name,
+		}, nil
```
```suggestion
		return addrs.NewDefaultProvider(provMeta.Name), nil
```
It will do the same thing but hide the "hack" in a nice official method :slightly_smiling_face:

Useurmind replied:
I will change it accordingly but keep the comment about the hack. Hiding hacks is bad. ;)

nhurel commented on lib/providers.go:
```diff
@@ -62,16 +67,98 @@ func (m *MockDataSourceReader) UnmatchedCalls() []cty.Value {
+	projectPluginDir := path.Join(dir, ".terraform/plugins/")
+	osArch := fmt.Sprintf("%s_%s", runtime.GOOS, runtime.GOARCH)
+	// TODO: this check could probably be improved
+	_, err := os.Stat(path.Join(projectPluginDir, osArch))
+	isTf13 := os.IsNotExist(err)
```
I wonder if we should add a cli flag so the user explicitely ask for a terraform 12 compatible behavior instead of having terraspec trying to guess in which condition it runs :thinking: 

Probably enough for now. I'll add a new flag in the future if it becomes too unpredictable

nhurel commented:
Last comments added on the PR. We're close to the end :slightly_smiling_face:

Useurmind commented:
Done with the comments above.
